libr/core/cmd_anal.c

```diff
@@ -1209,15 +1209,28 @@ static const char *syscallNumber(int n) {
 	return sdb_fmt (n > 1000 ? "0x%x" : "%d", n);
 }
 
-R_API char *cmd_syscall_dostr(RCore *core, int n) {
+const bool archIsThumb(RCore *core) {
+	return (!strcmp (core->anal->cur->arch, "arm") && core->anal->bits == 16);
+}
+
+R_API char *cmd_syscall_dostr(RCore *core, int n, ut64 addr) {
 	char *res = NULL;
 	int i;
 	char str[64];
 	if (n == -1) {
-		n = (int)r_debug_reg_get (core->dbg, "oeax");
-		if (!n || n == -1) {
-			const char *a0 = r_reg_get_name (core->anal->reg, R_REG_NAME_SN);
-			n = (int)r_debug_reg_get (core->dbg, a0);
+		if (addr && archIsThumb (core)) {
+			RAnalOp aop = {0};
+			ut8 buf[64];
+			r_io_read_at_mapped (core->io, addr, buf, sizeof(buf));
+			r_anal_op (core->anal, &aop, core->offset, buf, sizeof(buf), R_ANAL_OP_MASK_BASIC);
+			n = aop.val;
+			r_anal_op_fini (&aop);
```

Comment:
why not pass aop.val instead of the offset and avoid doing all this read+anal stuff in here? because you already have this aop.val in some caller places iirc